utilities/include/alps/numeric/tensors/tensor_base.hpp

```diff
@@ -207,17 +207,70 @@ namespace alps {
         template<typename T2, typename St, typename = std::enable_if<std::is_same<Container, storageType>::value, void >>
         tensor_base(tensor_base<T2, Dim, St> &&rhs) noexcept: storage_(rhs.storage()), shape_(rhs.shape()), acc_sizes_(rhs.acc_sizes()) {}
 
-        /// Different type assignment
-        template<typename T2, typename St>
-        tensor_base < T, Dim, Container > &operator=(const tensor_base < T2, Dim, St> &rhs){
-          assert(size()==rhs.size());
+        // Different type assignments
+
+        // assign when lhs and rhs containers are same
+        /// Copy assignment
+        tensor_base < T, Dim, Container > & operator=(const tensor_base < T, Dim, Container > &rhs) {
+          shape_ = rhs.shape();
+          acc_sizes_ = rhs.acc_sizes();
           storage_ = rhs.storage();
           return *this;
-        };
+        }
+        /// Move assignment
+        tensor_base < T, Dim, Container > & operator=(tensor_base < T, Dim, Container > &&rhs) {
+          shape_ = std::move(rhs.shape());
+          acc_sizes_ = std::move(rhs.acc_sizes());
+          storage_ = std::move(rhs.storage());
+          return *this;
+        }
+
+        // assign when lhs and rhs containers are different
+        /// Copy assignment
+        template<typename Container2>
+        typename std::enable_if<!std::is_same<Container2, Container>::value, tensor_base < T, Dim, Container > >::type & 
+        operator=(const tensor_base < T, Dim, Container2 > &rhs) {
+          shape_ = rhs.shape();
+          acc_sizes_ = rhs.acc_sizes();
+          reshape(shape_);
+          std::copy(rhs.storage().data(), rhs.storage().data() + rhs.storage().size(), storage_.data());
+          return *this;
+        }
+        /// Move assignment
+        template<typename Container2>
+        typename std::enable_if<!std::is_same<Container2, Container>::value, tensor_base < T, Dim, Container > >::type & 
+        operator=(tensor_base < T, Dim, Container2 > &&rhs) {
+          shape_ = rhs.shape();
+          acc_sizes_ = rhs.acc_sizes();
+          reshape(shape_);
+          std::copy(rhs.storage().data(), rhs.storage().data() + rhs.storage().size(), storage_.data());
+          return *this;
+        }
+
+        // assignment for different rhs value type
```

Comment:
Looks like the following cases can be covered by the ones above by adding the `T2` template parameter?

Reply:
For T != T2 I need to do deep copy instead of move.
The copy constructor indeed can be replaced.

Reply:
But for different containers you'll have to do a copy for move assignments anyways right?

Reply:
oh, I see what you mean. yes

Reply:
ok. done I suppose